src/main/java/za/ac/sun/cs/search/singleagent/Board.java

```diff
@@ -137,9 +144,45 @@ public String toString() {
         return outputBuilder.toString();
     }
 
-    public Integer getHeuristicCostEstimate(Board board) {
-        return 1;
+    public Board undoMove(Direction move) {
+
+        switch (move) {
+            case UP:
+                return makeMove(Direction.DOWN);
+            case DOWN:
+                return makeMove(Direction.UP);
+            case LEFT:
+                return makeMove(Direction.RIGHT);
+            case RIGHT:
+                return makeMove(Direction.LEFT);
+            default:
+                return null;
+        }
     }
 
-    public abstract Board makeMove(Direction move);
+    public Board makeMove(Direction move) {
```

Comment:
`makeMove` and `swapTiles` are now in the abstract class, just means we need to explicitly cast them to either an `ImplictBoard` or `ExplicitBoard` where appropriate.

Reply:
Sure.